daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Ref.scala

```diff
@@ -364,4 +365,40 @@ object Ref {
       assertRight(fromString(s))
   }
 
+  /** Package versions are non-empty strings consisting of segments of digits (without leading zeros)
+    *      separated by dots: "(0|[1-9][0-9]*)(\.(0|[1-9][0-9]*))*"
+    */
+  final case class PackageVersion private[data] (segments: ImmArray[Int])
+      extends Ordered[PackageVersion] {
+    override def toString: String = segments.toSeq.mkString(".")
+
+    override def compare(that: PackageVersion): Int = {
+      import scala.math.Ordering.Implicits._
+
+      implicitly[Ordering[Seq[Int]]].compare(segments.toSeq, that.segments.toSeq)
+    }
+  }
+
+  object PackageVersion {
+    private val MaxPackageVersionLength = 1024
+    final def fromString(s: String): Either[String, PackageVersion] = for {
```

Comment:
If we only use this for 1.16 packages that have this enforced by the compiler that is fine. If not then we need to think about a fallback. Maybe `ComparablePackageVersion`/`NonComparablePackageVersion` where the latter just stores a string (and gets rejected if used in a upgrading context).

Reply:
Why would we differentiate? The only difference to the status quo is the size limit enforcement, which we can drop anyway.

Reply:
I think there is an implicit limit already in the compiler. (packageName + packageVersion should be small than something). I have to double check with compiler guys.